workflow/sync/database_semaphore.go

```diff
@@ -0,0 +1,527 @@
+	for {
+		_, err := s.info.session.SQL().
+			DeleteFrom(s.info.config.lockTable).
+			Where(db.Cond{lockNameField: s.longDbKey()}).
+			Exec()
+		if err == nil {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+func (s *databaseSemaphore) release(key string) bool {
+	_, err := s.info.session.SQL().
+		DeleteFrom(s.info.config.stateTable).
+		Where(db.Cond{stateHeldField: true}).
+		And(db.Cond{stateNameField: s.longDbKey()}).
+		And(db.Cond{stateKeyField: key}).
+		And(db.Cond{stateControllerField: s.info.config.controllerName}).
+		Exec()
+
+	switch err {
+	case nil:
+		s.log.WithField("key", key).Debug("Released lock")
+		s.notifyWaiters()
+		return true
+	default:
+		s.log.WithField("key", key).WithError(err).Error("Failed to release lock")
+		return false
+	}
+}
+
+func (s *databaseSemaphore) queueOrdered(session db.Session) ([]stateRecord, error) {
+	since := time.Now().Add(-s.info.config.inactiveControllerTimeout)
+	var queue []stateRecord
+	subquery := session.SQL().
+		Select(controllerNameField).
+		From(s.info.config.controllerTable).
+		And(db.Cond{controllerTimeField + " >": since})
+
+	err := session.SQL().
+		Select(stateKeyField, stateControllerField).
+		From(s.info.config.stateTable).
+		Where(db.Cond{stateNameField: s.longDbKey()}).
+		And(db.Cond{stateHeldField: false}).
+		And(db.Cond{
+			"controller IN": subquery,
+		}).
+		OrderBy(statePriorityField+" DESC", stateTimeField+" ASC").
+		All(&queue)
+
+	if err != nil {
+		s.log.WithError(err).Error("Failed to get ordered queue for semaphore notification")
+		return nil, err
+	}
+	return queue, nil
+}
+
+// notifyWaiters enqueues the next N workflows who are waiting for the semaphore to the workqueue,
+// where N is the availability of the semaphore. If semaphore is out of capacity, this does nothing.
+func (s *databaseSemaphore) notifyWaiters() {
+	limit := s.getLimit()
+	// We don't need to run a transaction here, if we get it wrong it'll right itself
+	holders, err := s.getCurrentHolders()
+	if err != nil {
+		s.log.WithError(err).Error("Failed to notify waiters")
+		return
+	}
+	holdCount := len(holders)
+
+	pending, err := s.queueOrdered(s.info.session)
+	if err != nil {
+		return
+	}
+	triggerCount := min(limit-holdCount, len(pending))
+	s.log.WithFields(log.Fields{
+		"holdCount":    holdCount,
+		"triggerCount": triggerCount,
+		"pendingCount": len(pending),
+	}).Debug("Notifying waiters for semaphore")
+	for idx := 0; idx < triggerCount; idx++ {
+		item := pending[idx]
+		if item.Controller != s.info.config.controllerName {
+			continue
+		}
+		key := workflowKey(item.Key)
+		s.log.WithFields(log.Fields{"key": item.Key, "workflowKey": key}).Debug("Enqueueing workflow for semaphore notification")
+		s.nextWorkflow(key)
+	}
+}
+
+// addToQueue adds the holderkey into priority queue that maintains the priority order to acquire the lock.
+func (s *databaseSemaphore) addToQueue(holderKey string, priority int32, creationTime time.Time) {
```

Comment:
shouldn't the function signature here change to return an error? I believe this basically returns `true` in `prepAcquire` for example

Reply:
Yeah, not a lot we can do with the failure, but yes.